### DISCUSSION TITLE
Type inferrence modified in pinia 2.2.3, was working fine until then

### DISCUSSION THREAD

**Discussion by stephane303** [vuejs/pinia#2791] [Help and Questions]
### Reproduction

https://github.com/stephane303/pinia-demo

### Steps to reproduce the bug

To reproduce:

```sh
pnpm install

Check types:

npx vue-tsc --build --force

```

You will get
```sh
src/App.vue:6:19 - error TS7006: Parameter 'zone' implicitly has an 'any' type.

6 store.zones.find((zone) => zone.r =='b');
                    ~~~~


Found 1 error.
```

If you change the version of pinia to 2.2.2

You will get no error

### Expected behavior

Type inference is not modified

### Actual behavior

Type inference not working any more

**Comment by stephane303** [comment]
@posva 
This demo shows that the readonly solution, which was functioning correctly up until version 2.2.2, no longer works after the update to version 2.2.3. I believe this is an issue that needs to be addressed.

I have taken the time to prepare a demo that clearly demonstrates the problem introduced with the version change. However, instead of addressing the issue, it was moved to a discussion without any explanation. Could you please clarify the reasoning behind this?

**Reply by posva** [reply]
The repro is excellent! I appreciate it a lot, thanks! I gave it a look and it seems to work with the latest [versions](https://play.pinia.vuejs.org/#eNqNVE1v2zAM/SuCL3GARF7RoQO8OGg39LAVWIdmp8U5ODaTqrUlQ6LTZIH/+yj5o85SFDsEkclH6r0nUUfvpiz5rgIv9GYm1aJEZgCrkuWJ3Eaxhyb25rEURak0siOrDCwAUcitWaDSwGq20apgIx6kqpIIehTLWKZKGurkENFZkT/+bEEuzf8oCYZvhMx8367HLJozu+CaRdFoPWrAs6ChR2ToA6Eo8wSBvhizPwSDFtQnvInXkJ4WScmfjJIk8ejgbYKUhcxFbKwUUiQ2EnuBW3MwxXSt1YsBTfWxN+mg1+RXkMEOlcrNNClFU/aIWJowCNJMEj6DXOw0l4CBLIvgrOb6il/xiyAX64A2Ckg+7E+3oYppBsX/dO+g1x/4xUf+yXXdXbZNC9vVNq1jWZMtaOh0NmL7jympKkqRg74vUdDpnZiT5Ll6+e5iqCvoKaaPkD6/ETeoRYpn4Sezb8T81ECu7mCgFhO9BVcSe7eLH7CndZ8sVFblhH4n+QBG5ZWl3sC+VDIjNQOcE/HNHT3dxF/mdo8gTafVEnUmObxz/+s7jrzSveSXA3PbIeBoyNl+ajKg+w2nA+MuGQ1LD7IHUCFkE6YhSVHsYMIsBRt7gE1fSMzcjOGhBPabBoUm7Mh0SOOmSZhlEkvYu67CktkkKbAF0lR0/N18mZDdaJ0cZrbHnIZsWNgM8Nm0R0MpPh10myWbme8mt92hewBo07DdO+p1zVxg7ndsXgktrZBRMqon7EiL9aheNZi6eQVeW1OvTMn80PW+Xz9BitxadCvJCDB+G3qGg/EdkzGnt8D3KeCoLmkx6W33G/4OaDMsMYz+1IZZo+nPZVbj1XhsUwPuyzvy+S1sODy/RvTybmWtdpJ6RZpeXC2HLTnnJwLbS2xLyAmv/gvvROrF).

BTW, in your demo you are not returning the original state. Maybe this is just for the demo but FYI this will fail for devtools, SSR, and some pinia plugins because they won't be able to read the state. I wrote about ways to handle this in this [article](https://masteringpinia.com/blog/how-to-create-private-state-in-stores) and cover it deeply in the Mastering Pinia course itself.

To answer your question about the move to discussions, I have my way of handling issues in Pinia to help me organize my life. That's the reasoning without going too deep. In this case

**Reply by stephane303** [reply]
Yes, it is functioning, but it no longer passes the type-check test as of version 2.2.3:

```
npx vue-tsc --build --force
```
This is the root of the issue.

Regarding your article, I am not entirely sure I understand your point, and I don't believe it's necessary to pay for access to basic functionality instructions. What I require is read-only access to the state for the reasons mentioned [here](https://github.com/vuejs/pinia/discussions/2792).

**Reply by posva** [reply]
By working, I mean the type check. The playground also has TypeScript checks but it uses up to date versions of everything.

> Regarding your article, I am not entirely sure I understand your point, and I don't believe it's necessary to pay for access to basic functionality instructions. 

The article is completely free... The rest of the course is far from being _"basic functionality instructions"_. I'm just trying to point you to content that could help.

**Reply by stephane303** [reply]
I am using a CI, and I need to do 
```
npx vue-tsc --build --force
```

This is broken since version 2.2.3

To me having a readonly state is basic functionality.

**Reply by vaasmundsen** [reply]
I am running into the same issue in Vue 2.7, but it is working as expected in Vue 3.5.

In Vue 2.7
```ts
const { zones } = storeToRefs(useSettingsStore());
// ^^ const zones: WritableComputedRef<any>
```

In Vue 3.5
```ts
const { zones } = storeToRefs(useSettingsStore());
// ^^ const zones: ComputedRef<Zone[]>
```

If I modify the store code to return a WritableComputedRef, it will infer it correct Vue 2.7 as well.
```ts
// Change to writeable computed
  const readonlyState = Object.fromEntries(
    Object.keys(state).map((key) => [
      key,
      computed({
        get() {
          return state[key as keyof typeof state];
        },
        set() {},
      }),
    ])
  );


const { zones } = storeToRefs(useSettingsStore());
// ^^ const zones: WritableComputedRef<Zone[]>
```

I am using the same typescript and vue-tsc versions in both projects.

```json
  "dependencies": {
    "pinia": "^2.2.4",
  },
  "devDependencies": {
    "typescript": "^5.6.3",
    "vite": "^5.4.9",
    "vue-tsc": "^2.1.6"
  }
```

If I change back to version 2.2.2 as mentioned by @stephane303 , it is inferred correctly for Vue 2.7 as well.

**Comment by stephane303** [comment]
[Why Readonly State Is Essential in Pinia](https://github.com/vuejs/pinia/discussions/2792)
